File: shapeshifter-py/shapeshifter/parser.py
"""
Shapeshifter parser: .ss source -> AST.

Implements the lexical structure and grammar of the language
specification (sections 3 and 4):

  - vacuous lines (blank / comment-only) are discarded  [Def. 3.2]
  - comment stripping precedes vacuity testing          [Rem. 3.3]
  - bracket-balance continuation joins logical lines    [Def. 3.5]
  - six block forms, field blocks vs statement blocks   [Def. 4.1, 4.2]
  - all call arguments are named                        [Def. 4.3]
  - value language with depth-zero comma splitting      [Def. 4.4, Lem. 4.6]
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any

OPEN, CLOSE = "[({", "])}"
QUOTES = ('"', "'")

FIELD_BLOCKS = ("objective", "instrument", "dataset", "target_list")
STMT_BLOCKS = ("phase", "validate")
LADDER_BLOCKS = ("ladder",)


class ParseError(Exception):
    """Raised when the source cannot be parsed; the compile stage reports it."""


# ---------------------------------------------------------------- values

def _split_depth0(raw: str) -> list[str]:
    """Split on commas at bracket depth zero.  [Lemma 4.6]"""
    parts, depth, start = [], 0, 0
    for i, c in enumerate(raw):
        if c in OPEN:
            depth += 1
        elif c in CLOSE:
            depth -= 1
        elif c == "," and depth == 0:
            parts.append(raw[start:i])
            start = i + 1
    parts.append(raw[start:])
    return [p for p in parts if p.strip()]


def _parse_object(raw: str) -> dict:
    inner = raw.strip()
    if inner.startswith("{"):
        inner = inner[1:]
    if inner.endswith("}"):
        inner = inner[:-1]
    obj: dict[str, Any] = {}
    for pair in _split_depth0(inner):
        ci = pair.find(":")
        if ci >= 0:
            obj[pair[:ci].strip()] = parse_value(pair[ci + 1:].strip())
    return obj


def _parse_object_array(raw: str) -> list[dict]:
    objs, depth, start = [], 0, 0
    for i, c in enumerate(raw):
        if c == "{":
            if depth == 0:
                start = i
            depth += 1
        elif c == "}":
            depth -= 1
            if depth == 0:
                objs.append(_parse_object(raw[start:i + 1]))
    return objs


def parse_value(raw: Any) -> Any:
    """Value parsing function P.  [Def. 4.5, clauses V1-V7]"""
    if raw is None:
        return None
    s = str(raw).strip()
    if not s:                                                    # V1
        return None
    if len(s) >= 2 and s[0] == s[-1] and s[0] in QUOTES:         # V2
        return s[1:-1]
    if s == "true":                                              # V3
        return True
    if s == "false":
        return False
    if s.startswith("[") and s.endswith("]"):                    # V4
        inner = s[1:-1].strip()
        if not inner:
            return []
        if inner.startswith("{"):
            return _parse_object_array(inner)
        return [parse_value(v) for v in _split_depth0(inner)]
    if s.startswith("{") and s.endswith("}"):                    # V5
        return _parse_object(s)
    if re.fullmatch(r"[+-]?\d+", s):                             # V6
        return int(s)
    try:
        return float(s)
    except ValueError:
        pass
    return s                                                     # V7 bare word


# ---------------------------------------------------------------- AST

@dataclass
class Call:
    fn: str
    args: dict[str, Any]


@dataclass
class Assign:
    target: str
    value: Any            # Call when is_call, else a literal
    is_call: bool


@dataclass
class Rung:
    """One contact. `power` is a resolution increment in [0,1), not a
    displacement --- see Definition 3.2 of the instrument paper."""
    name: str
    power: float


@dataclass
class Ladder:
    """An ordered contact sequence with an optional declared target."""
    name: str
    toward: str | None
    rungs: list = field(default_factory=list)
    require: dict | None = None      # {"metric":..., "op":..., "value":...}


@dataclass
class AST:
    imports: list[str] = field(default_factory=list)
    objective: dict | None = None
    instruments: dict[str, dict] = field(default_factory=dict)
    datasets: dict[str, dict] = field(default_factory=dict)
    target_lists: dict[str, dict] = field(default_factory=dict)
    phases: dict[str, list] = field(default_factory=dict)
    validates: dict[str, list] = field(default_factory=dict)
    ladders: dict[str, Ladder] = field(default_factory=dict)

    def to_dict(self) -> dict:
        def stmt(s):
            if isinstance(s, Assign):
                v = ({"type": "call", "fn": s.value.fn, "args": s.value.args}
                     if s.is_call else {"type": "value", "value": s.value})
                return {"type": "assign", "target": s.target, "value": v}
            return {"type": "call", "fn": s.fn, "args": s.args}

        return {
            "imports": self.imports,
            "objective": self.objective,
            "instruments": self.instruments,
            "datasets": self.datasets,
            "target_lists": self.target_lists,
            "phases": {k: [stmt(s) for s in v] for k, v in self.phases.items()},
            "validates": {k: [stmt(s) for s in v]
                          for k, v in self.validates.items()},
            "ladders": {
                k: {"name": L.name, "toward": L.toward,
                    "rungs": [{"name": r.name, "power": r.power}
                              for r in L.rungs],
                    "require": L.require}
                for k, L in self.ladders.items()
            },
        }


# ---------------------------------------------------------------- lines

@dataclass
class Line:
    num: int
    indent: int
    body: str


def _logical_lines(source: str) -> list[Line]:
    """Strip comments, drop vacuous lines, join by bracket balance."""
    raw: list[Line] = []
    for i, ln in enumerate(source.split("\n")):
        body = re.sub(r"//.*$", "", ln).strip()
        if body:
            raw.append(Line(i + 1, len(ln) - len(ln.lstrip()), body))

    def bal(s: str) -> int:
        return sum(c in OPEN for c in s) - sum(c in CLOSE for c in s)

    out: list[Line] = []
    i = 0
    while i < len(raw):
        cur = raw[i]
        b = bal(cur.body)
        if b > 0:
            combined, j = cur.body, i + 1
            while j < len(raw) and b > 0:
                combined += " " + raw[j].body
                b += bal(raw[j].body)
                j += 1
            out.append(Line(cur.num, cur.indent, combined))
            i = j
        else:
            out.append(cur)
            i += 1
    return out


def _fields(lines: list[Line], i: int, base: int) -> tuple[dict, int]:
    out: dict[str, Any] = {}
    while i < len(lines) and lines[i].indent > base:
        m = re.match(r"^(\w+)\s*:\s*(.*)$", lines[i].body)
        if m:
            out[m.group(1)] = parse_value(m.group(2))
        i += 1
    return out, i


def _args(raw: str) -> dict:
    """All arguments are named.  [Def. 4.3, Rem. 4.7]"""
    args: dict[str, Any] = {}
    for part in _split_depth0(raw):
        ci = part.find(":")
        if ci >= 0:
            args[part[:ci].strip()] = parse_value(part[ci + 1:].strip())
    return args


RUNG_RE = re.compile(r"^rung\s+(\w+)\s+at\s+([-+0-9.eE]+)$")
REQUIRE_RE = re.compile(r"^require\s+(\w+)\s*(>=|<=|==|>|<)\s*([-+0-9.eE]+)$")


def _ladder_body(lines: list[Line], i: int, base: int,
                 name: str, toward: str | None) -> tuple[Ladder, int]:
    """Read a ladder block.

    Two statement forms only:
        rung <name> at <power>
        require <metric> <op> <value>

    A rung's power is a resolution increment; the parser enforces the
    half-open range [0,1) because a rung of resolution 1 would collapse
    the ambiguity to the floor in one contact, which Lemma 3.3 excludes.
    """
    lad = Ladder(name=name, toward=toward)
    while i < len(lines) and lines[i].indent > base:
        t = lines[i].body
        m = RUNG_RE.match(t)
        if m:
            power = float(m.group(2))
            if not (0.0 <= power < 1.0):
                raise ParseError(
                    f"line {lines[i].num}: rung {m.group(1)!r} has resolution "
                    f"{power}, outside [0,1)")
            lad.rungs.append(Rung(m.group(1), power))
            i += 1
            continue
        m = REQUIRE_RE.match(t)
        if m:
            lad.require = {"metric": m.group(1), "op": m.group(2),
                           "value": float(m.group(3))}
            i += 1
            continue
        raise ParseError(f"line {lines[i].num}: unrecognised ladder "
                         f"statement {t!r}")
    return lad, i


def _statements(lines: list[Line], i: int, base: int) -> tuple[list, int]:
    out: list[Any] = []
    while i < len(lines) and lines[i].indent > base:
        t = lines[i].body
        m = re.match(r"^(\w+)\s*=\s*(.+)$", t, re.S)
        if m:
            target, expr = m.group(1), m.group(2).strip()
            c = re.match(r"^([\w.]+)\s*\((.*)\)$", expr, re.S)
            if c:
                out.append(Assign(target, Call(c.group(1), _args(c.group(2))), True))
            else:
                out.append(Assign(target, parse_value(expr), False))
        else:
            c = re.match(r"^([\w.]+)\s*\((.*)\)$", t, re.S)
            if c:
                out.append(Call(c.group(1), _args(c.group(2))))
        i += 1
    return out, i


def parse(source: str) -> AST:
    """Parse .ss source into an AST."""
    lines = _logical_lines(source)
    ast, i = AST(), 0
    while i < len(lines):
        ln = lines[i]

        if ln.body.startswith("import "):
            ast.imports.append(ln.body[7:].strip())
            i += 1
            continue

        lm = re.match(r"^ladder\s+(\w+)(?:\s+toward\s+(\w+))?\s*:?$",
                      ln.body)
        if lm:
            lad, i = _ladder_body(lines, i + 1, ln.indent,
                                  lm.group(1), lm.group(2))
            ast.ladders[lad.name] = lad
            continue

        m = re.match(r"^(\w+)\s+(\w+)\s*:$", ln.body)
        if m and m.group(1) in FIELD_BLOCKS:
            kw, name = m.group(1), m.group(2)
            body, i = _fields(lines, i + 1, ln.indent)
            if kw == "objective":
                ast.objective = {"name": name, "fields": body}
            elif kw == "instrument":
                ast.instruments[name] = body
            elif kw == "dataset":
                ast.datasets[name] = body
            else:
                ast.target_lists[name] = {"name": name, **body}
            continue

        if m and m.group(1) in STMT_BLOCKS:
            kw, name = m.group(1), m.group(2)
            body, i = _statements(lines, i + 1, ln.indent)
            (ast.phases if kw == "phase" else ast.validates)[name] = body
            continue

        i += 1
    return ast

File: catalogue/web/src/lib/ss/parser.js
// =====================================================================
//  Shapeshifter parser --- a literal transcription of
//  shapeshifter-py/shapeshifter/parser.py
//
//  This is a PORT, not a reimplementation. Where the Python does
//  something surprising, so does this file, and a comment says why.
//  The oracle is shapeshifter-py/results/*.json; scripts/check_port.mjs
//  diffs against it.
// =====================================================================

const OPEN = '[({'
const CLOSE = '])}'
const QUOTES = ['"', "'"]

export const FIELD_BLOCKS = ['objective', 'instrument', 'dataset', 'target_list']
export const STMT_BLOCKS = ['phase', 'validate']

export class ParseError extends Error {
  constructor(message) {
    super(message)
    this.name = 'ParseError'
  }
}

// ------------------------------------------------------------- values

/* Split on commas at bracket depth zero.

   The Python counts brackets and ignores quotes entirely, so a comma
   inside a string literal at depth zero splits the value. That is a bug,
   and it is preserved here: a port that fixed it would accept programs
   the reference implementation rejects. */
export function splitDepth0(raw) {
  const parts = []
  let depth = 0
  let start = 0
  for (let i = 0; i < raw.length; i++) {
    const c = raw[i]
    if (OPEN.includes(c)) depth += 1
    else if (CLOSE.includes(c)) depth -= 1
    else if (c === ',' && depth === 0) {
      parts.push(raw.slice(start, i))
      start = i + 1
    }
  }
  parts.push(raw.slice(start))
  return parts.filter((p) => p.trim())
}

function parseObject(raw) {
  let inner = raw.trim()
  if (inner.startsWith('{')) inner = inner.slice(1)
  if (inner.endsWith('}')) inner = inner.slice(0, -1)
  const obj = {}
  for (const pair of splitDepth0(inner)) {
    const ci = pair.indexOf(':')
    if (ci >= 0) obj[pair.slice(0, ci).trim()] = parseValue(pair.slice(ci + 1).trim())
  }
  return obj
}

function parseObjectArray(raw) {
  const objs = []
  let depth = 0
  let start = 0
  for (let i = 0; i < raw.length; i++) {
    const c = raw[i]
    if (c === '{') {
      if (depth === 0) start = i
      depth += 1
    } else if (c === '}') {
      depth -= 1
      if (depth === 0) objs.push(parseObject(raw.slice(start, i + 1)))
    }
  }
  return objs
}

/* Python float() and JS parseFloat() disagree at both ends: float() takes
   "nan" / "inf" / "1_000", parseFloat takes "1abc". Neither loose form is
   admitted, so this regex is the intersection --- what a Python float
   literal looks like and nothing else. Anything outside it falls through
   to V7 and stays a bare word, exactly as the Python does. */
const FLOAT_RE = /^[+-]?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$/
const INT_RE = /^[+-]?\d+$/

// Value parsing, clauses V1-V7 in the Python order.
export function parseValue(raw) {
  if (raw === null || raw === undefined) return null
  const s = String(raw).trim()
  if (!s) return null                                              // V1
  if (s.length >= 2 && s[0] === s[s.length - 1] && QUOTES.includes(s[0]))
    return s.slice(1, -1)                                          // V2 no escapes
  if (s === 'true') return true                                    // V3
  if (s === 'false') return false
  if (s.startsWith('[') && s.endsWith(']')) {                      // V4
    const inner = s.slice(1, -1).trim()
    if (!inner) return []
    if (inner.startsWith('{')) return parseObjectArray(inner)
    return splitDepth0(inner).map(parseValue)
  }
  if (s.startsWith('{') && s.endsWith('}')) return parseObject(s)   // V5
  if (INT_RE.test(s)) return parseInt(s, 10)                       // V6
  if (FLOAT_RE.test(s)) {
    const f = Number(s)
    if (Number.isFinite(f)) return f
  }
  return s                                                         // V7 bare word
}

// ------------------------------------------------------------- lines

function balance(s) {
  let b = 0
  for (const c of s) {
    if (OPEN.includes(c)) b += 1
    else if (CLOSE.includes(c)) b -= 1
  }
  return b
}

/* Strip comments, drop vacuous lines, join by bracket balance.

   Order matters: the comment is removed BEFORE the vacuity test, so a
   comment-only line disappears entirely. The indent is measured on the
   ORIGINAL raw line, not on the stripped body. A continued line keeps the
   first line number and the first indent, and the pieces join with a
   single space. */
function logicalLines(source) {
  const raw = []
  const src = source.split('\n')
  for (let i = 0; i < src.length; i++) {
    const ln = src[i].replace(/\r$/, '')
    const body = ln.replace(/\/\/.*$/, '').trim()
    if (body) raw.push({ num: i + 1, indent: ln.length - ln.replace(/^\s+/, '').length, body })
  }

  const out = []
  let i = 0
  while (i < raw.length) {
    const cur = raw[i]
    let b = balance(cur.body)
    if (b > 0) {
      let combined = cur.body
      let j = i + 1
      while (j < raw.length && b > 0) {
        combined += ' ' + raw[j].body
        b += balance(raw[j].body)
        j += 1
      }
      out.push({ num: cur.num, indent: cur.indent, body: combined })
      i = j
    } else {
      out.push(cur)
      i += 1
    }
  }
  return out
}

function readFields(lines, i, base) {
  const out = new Map()
  while (i < lines.length && lines[i].indent > base) {
    const m = /^(\w+)\s*:\s*([\s\S]*)$/.exec(lines[i].body)
    if (m) out.set(m[1], parseValue(m[2]))
    i += 1
  }
  return [out, i]
}

// Every argument is named; positional arguments cannot be expressed.
function readArgs(raw) {
  const args = new Map()
  for (const part of splitDepth0(raw)) {
    const ci = part.indexOf(':')
    if (ci >= 0) args.set(part.slice(0, ci).trim(), parseValue(part.slice(ci + 1).trim()))
  }
  return args
}

const RUNG_RE = /^rung\s+(\w+)\s+at\s+([-+0-9.eE]+)$/
const REQUIRE_RE = /^require\s+(\w+)\s*(>=|<=|==|>|<)\s*([-+0-9.eE]+)$/

/* Read a ladder block. Two statement forms only:

       rung <name> at <power>
       require <metric> <op> <value>

   A rung power is a resolution increment and is hard-validated into
   [0,1): a rung of resolution 1 would collapse the ambiguity to the floor
   in a single contact, which the non-completability axiom excludes.

   Ladders are the ONLY strict block --- an unrecognised statement raises.
   Every other block silently drops what it cannot match. */
function readLadderBody(lines, i, base, name, toward) {
  const lad = { name, toward: toward === undefined ? null : toward, rungs: [], require: null }
  while (i < lines.length && lines[i].indent > base) {
    const t = lines[i].body
    let m = RUNG_RE.exec(t)
    if (m) {
      const power = Number(m[2])
      if (!(power >= 0.0 && power < 1.0))
        throw new ParseError(
          'line ' + lines[i].num + ': rung ' + JSON.stringify(m[1]) +
            ' has resolution ' + power + ', outside [0,1)'
        )
      lad.rungs.push({ name: m[1], power })
      i += 1
      continue
    }
    m = REQUIRE_RE.exec(t)
    if (m) {
      // A second require overwrites the first; only one survives.
      lad.require = { metric: m[1], op: m[2], value: Number(m[3]) }
      i += 1
      continue
    }
    throw new ParseError(
      'line ' + lines[i].num + ': unrecognised ladder statement ' + JSON.stringify(t)
    )
  }
  return [lad, i]
}

function readStatements(lines, i, base) {
  const out = []
  while (i < lines.length && lines[i].indent > base) {
    const t = lines[i].body
    const m = /^(\w+)\s*=\s*([\s\S]+)$/.exec(t)
    if (m) {
      const target = m[1]
      const expr = m[2].trim()
      const c = /^([\w.]+)\s*\(([\s\S]*)\)$/.exec(expr)
      if (c)
        out.push({
          type: 'assign', target, isCall: true,
          call: { fn: c[1], args: readArgs(c[2]) },
        })
      else out.push({ type: 'assign', target, isCall: false, value: parseValue(expr) })
    } else {
      const c = /^([\w.]+)\s*\(([\s\S]*)\)$/.exec(t)
      // A bare call parses and is audited at compile time, but
      // execute_stage skips it: only assignments ever run. Preserved.
      if (c) out.push({ type: 'call', fn: c[1], args: readArgs(c[2]) })
    }
    i += 1
  }
  return [out, i]
}

// ------------------------------------------------------------- parse

/* Blocks live in Maps keyed by name: a redeclaration overwrites, and
   insertion order is execution order. */
export function parse(source) {
  const lines = logicalLines(source)
  const ast = {
    imports: [],
    objective: null,
    instruments: new Map(),
    datasets: new Map(),
    target_lists: new Map(),
    phases: new Map(),
    validates: new Map(),
    ladders: new Map(),
  }
  let i = 0
  while (i < lines.length) {
    const ln = lines[i]

    if (ln.body.startsWith('import ')) {
      ast.imports.push(ln.body.slice(7).trim())
      i += 1
      continue
    }

    // The ladder form is tested BEFORE the generic block form, and it is
    // the only block whose trailing colon is optional.
    const lm = /^ladder\s+(\w+)(?:\s+toward\s+(\w+))?\s*:?$/.exec(ln.body)
    if (lm) {
      const [lad, ni] = readLadderBody(lines, i + 1, ln.indent, lm[1], lm[2])
      ast.ladders.set(lad.name, lad)
      i = ni
      continue
    }

    const m = /^(\w+)\s+(\w+)\s*:$/.exec(ln.body)
    if (m && FIELD_BLOCKS.includes(m[1])) {
      const kw = m[1]
      const name = m[2]
      const [body, ni] = readFields(lines, i + 1, ln.indent)
      i = ni
      if (kw === 'objective') ast.objective = { name, fields: body }
      else if (kw === 'instrument') ast.instruments.set(name, body)
      else if (kw === 'dataset') ast.datasets.set(name, body)
      else {
        // target_list merges its own name INTO the field map; objective
        // keeps the name outside, under a separate key.
        const merged = new Map([['name', name]])
        for (const [k, v] of body) merged.set(k, v)
        ast.target_lists.set(name, merged)
      }
      continue
    }

    if (m && STMT_BLOCKS.includes(m[1])) {
      const kw = m[1]
      const name = m[2]
      const [body, ni] = readStatements(lines, i + 1, ln.indent)
      i = ni
      ;(kw === 'phase' ? ast.phases : ast.validates).set(name, body)
      continue
    }

    i += 1 // anything unmatched is silently skipped
  }
  return ast
}

const mapToObj = (m) => Object.fromEntries(m)

// Mirrors AST.to_dict(); this shape is what becomes compile.ir.
export function astToDict(ast) {
  const stmt = (s) => {
    if (s.type === 'assign') {
      const v = s.isCall
        ? { type: 'call', fn: s.call.fn, args: mapToObj(s.call.args) }
        : { type: 'value', value: s.value }
      return { type: 'assign', target: s.target, value: v }
    }
    return { type: 'call', fn: s.fn, args: mapToObj(s.args) }
  }
  const stmtMap = (m) => Object.fromEntries([...m].map(([k, v]) => [k, v.map(stmt)]))

  return {
    imports: ast.imports,
    objective: ast.objective
      ? { name: ast.objective.name, fields: mapToObj(ast.objective.fields) }
      : null,
    instruments: Object.fromEntries([...ast.instruments].map(([k, v]) => [k, mapToObj(v)])),
    datasets: Object.fromEntries([...ast.datasets].map(([k, v]) => [k, mapToObj(v)])),
    target_lists: Object.fromEntries([...ast.target_lists].map(([k, v]) => [k, mapToObj(v)])),
    phases: stmtMap(ast.phases),
    validates: stmtMap(ast.validates),
    ladders: Object.fromEntries(
      [...ast.ladders].map(([k, L]) => [
        k,
        {
          name: L.name,
          toward: L.toward,
          rungs: L.rungs.map((r) => ({ name: r.name, power: r.power })),
          require: L.require,
        },
      ])
    ),
  }
}
